{{cookiecutter.project_slug}}/server/{{cookiecutter.project_slug}}/utils/tests.py

```diff
@@ -0,0 +1,32 @@
+import pytest
+
+from {{ cookiecutter.project_slug }}.utils.sites import get_site_url
+
+
+@pytest.mark.parametrize(
+    "custom_settings,expected_output",
+    [
+        ({"CURRENT_DOMAIN": "localhost", "CURRENT_PORT": 8080, "HEROKU_APP_NAME": None, "IN_DEV": True}, "http://localhost:8080"),
+        ({"CURRENT_DOMAIN": "http://localhost/", "CURRENT_PORT": None, "HEROKU_APP_NAME": "", "IN_DEV": False}, "https://localhost"),
+        ({"CURRENT_DOMAIN": "", "CURRENT_PORT": None, "HEROKU_APP_NAME": "example", "IN_DEV": False}, "https://example.herokuapp.com"),
+    ],
+)
+def test_get_site_url(settings, custom_settings, expected_output):
+    for key in custom_settings:
+        settings.__setattr__(key, custom_settings[key])
+    site_url = get_site_url()
+    assert site_url == expected_output
+
+
+@pytest.mark.parametrize(
+    "custom_settings,expected_output",
+    [
+        ({"CURRENT_DOMAIN": None, "CURRENT_PORT": 8080, "HEROKU_APP_NAME": None, "IN_DEV": True}, "http://localhost:8080"),
+        ({"CURRENT_DOMAIN": "", "CURRENT_PORT": None, "HEROKU_APP_NAME": "", "IN_DEV": False}, "https://example.herokuapp.com"),
+    ],
+)
+def test_get_site_url_negative(settings, custom_settings, expected_output):
+    for key in custom_settings:
+        settings.__setattr__(key, custom_settings[key])
+    with pytest.raises(Exception):
+        get_site_url()
```

Comment:
great question! This is one of the oddities of pytest. In this case specifically `pytest-django` is injecting this fixture and making it available for all tests. So for any test to have access to the loaded settings or to manipulate them, you simple add it to the test like I did here.
And `parameterize` has the same oddity. You just have to KNOW that those key names will be the last X number of arguments for the function and that any other fixtures you are using have to come before.
I've been in situations where I'm doing what I've done here...but ALSO using several other fixtures, and it can get kludgy and confusing very quickly.

Reply:
I see - and is the ability to use `settings` as a param "built in" to pytest, or is it configured in the code somewhere?

Reply:
built-in to `pytest-django` I believe. So it's a bit of magic that you wouldn't know about unless you read the docs or inspect all the fixtures that are initialized (which you can do, I forget how though)